source/slang/slang-ir-use-uninitialized-values.cpp

```diff
@@ -348,8 +348,135 @@ namespace Slang
         return loads;
     }
 
+    static bool isInstStoredInto(ReachabilityContext& reachability, IRInst* reference, IRInst* inst)
+    {
+        auto addresses = getAliasableInstructions(inst);
+        
+        List<IRInst*> stores;
+        List<IRInst*> loads;
+
+        for (auto alias : addresses)
+        {
+            for (auto use = alias->firstUse; use; use = use->nextUse)
+            {
+                IRInst* user = use->getUser();
+                collectLoadStore(stores, loads, user, alias);
+            }
+        }
+
+        for (auto store : stores)
+        {
+            if (reachability.isInstReachable(store, reference))
+                return true;
+        }
+
+        return false;
+    }
+
+    static IRInst* traceInstOrigin(IRInst* inst)
+    {
+        if (auto load = as<IRLoad>(inst))
+            return traceInstOrigin(load->getPtr());
+
+        return inst;
+    }
+
+    static bool isFieldUsed(IRStructField* field, const List<IRStructKey*>& used)
+    {
+        for (auto key : used)
+        {
+            if (field->getKey() == key)
+                return true;
+        }
+
+        return false;
+    }
+
+    static List<IRStructField*> checkFieldsFromExit(ReachabilityContext& reachability, IRReturn* ret, IRStructType* type)
+    {
+        IRInst* origin = traceInstOrigin(ret->getVal());
+
+        // We don't want to warn on delegated construction
+        if (!isUninitializedValue(origin))
+            return {};
+
+        // Now we can look for all references to fields
+        List<IRStructKey*> usedKeys;
+        for (auto use = origin->firstUse; use; use = use->nextUse)
+        {
+            IRInst* user = use->getUser();
+            
+            auto fieldAddress = as<IRFieldAddress>(user);
+            if (!fieldAddress || !isInstStoredInto(reachability, ret, user))
+                continue;
+
+            IRInst* field = fieldAddress->getField();
+            usedKeys.add(as<IRStructKey>(field));
+        }
+
+        List<IRStructField*> uninitializedFields;
+
+        auto fields = type->getFields();
+        for (auto field : fields)
+        {
+            if (canIgnoreType(field->getFieldType(), nullptr))
+                continue;
+
+            if (!isFieldUsed(field, usedKeys))
+                uninitializedFields.add(field);
+        }
+        
+        return uninitializedFields;
+    }
+
+    static void checkConstructor(IRFunc* func, ReachabilityContext& reachability, DiagnosticSink* sink)
+    {
+        IRStructType* stype = as<IRStructType>(func->getResultType());
+        if (!stype)
+            return;
+        
+        auto synthesized = func->findDecoration<IRSynthesizedDecoration>();
+        
+        auto emitWarnings = [&](const List<IRStructField*>& fields, IRReturn* ret)
```

Comment:
I don't think we want to use `emit` for printing warnings.
You may want to reword it to `printWarning`, `printDiagonostics` or something like that.

Reply:
Sounds good. Is lambda use alright?

Reply:
We have a rule for lambda.
```
Lambdas are allowed, but think carefully about whether just declaring a subroutine would also work.
```
https://github.com/jkwak-work/slang/blob/master/docs/design/coding-conventions.md
In general, it is better to stay away from it.
In this case, it is called only once, my preference is not to use it, but it seems fine.

Reply:
I see. In this case I wanted to avoid passing multiple parameters to a separate function.